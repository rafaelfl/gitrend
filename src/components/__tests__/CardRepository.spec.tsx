import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { CardRepository } from '..';

describe('CardRepository component tests', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    const buildCardRepository = (
        title: string,
        description: string,
        language: string,
        stars: number,
        fork: number,
        isFavorite: boolean,
        url: string,
        onFavoriteClick?: () => void,
    ) => (
        <CardRepository
            title={title}
            description={description}
            language={language}
            stars={stars}
            forks={fork}
            isFavorite={isFavorite}
            onFavoriteClick={onFavoriteClick}
            url={url}
        />
    );

    test('it should match snapshot', () => {
        const { container } = render(
            buildCardRepository(
                'microsoft / PowerToys',
                'Small educational project developed with the objective of presenting the concepts related to the development of backend REST APIs with Node.js for the course Distributed Systems (2021.2) of the Computer Engineering Course - Federal University of Maranhão (UFMA)(UFMA)(UFMA)(UFMA)(UFMA)(UFMA)(UFMA)(UFMA)(UFMA)(UFMA)(UFMA)',
                'C#',
                67934,
                3816,
                true,
                'https://github.com/microsoft/terminal',
            ),
        );

        expect(container).toMatchSnapshot();
    });

    test('it should render the card with all info and truncated description', () => {
        const descText =
            'Small educational project developed with the objective of presenting the concepts related to the development of backend REST APIs with Node.js for the course Distributed Systems (2021.2) of the Computer Engineering Course - Federal University of Maranhão (UFMA)';

        const { getByText, getByRole } = render(
            buildCardRepository(
                'Microsoft / PowerToys',
                descText,
                'C#',
                67934,
                3816,
                true,
                'https://github.com/microsoft/terminal',
            ),
        );

        expect(getByText(/Microsoft/)).toBeInTheDocument();
        expect(getByText(/C#/)).toBeInTheDocument();
        expect(getByText(/67,934/)).toBeInTheDocument();
        expect(getByText(/3,816/)).toBeInTheDocument();
        expect(getByText('https://github.com/microsoft/terminal')).toBeInTheDocument();

        const descriptionDiv = getByText(/Small educational/);
        expect(descriptionDiv).toBeInTheDocument();

        // the description should be truncated
        const descriptionText = descriptionDiv.textContent;

        expect(descriptionText).not.toEqual(descText);
        expect(descriptionText).toHaveLength(250);

        const favoriteButton = getByRole('button');
        expect(favoriteButton).toBeInTheDocument();

        const favoriteIcon = favoriteButton.firstChild;
        expect(favoriteIcon).toHaveClass('fas', 'fa-heart', 'fa-lg');
        expect(favoriteIcon).not.toHaveClass('far');
    });

    test('it should change the favorite icon when the favorite button is clicked', () => {
        let favorite = false;

        const onFavoriteClick = jest.fn(() => (favorite = !favorite));

        const { getByRole, rerender } = render(
            buildCardRepository(
                'PowerToys',
                'Small educational project',
                'C#',
                67934,
                3816,
                favorite,
                'https://github.com/microsoft/terminal',
                onFavoriteClick,
            ),
        );

        const favoriteButton = getByRole('button');
        expect(favoriteButton).toBeInTheDocument();

        const favoriteIcon = favoriteButton.firstChild;
        expect(favoriteIcon).toHaveClass('far', 'fa-heart', 'fa-lg');
        expect(favoriteIcon).not.toHaveClass('fas');

        userEvent.click(favoriteButton);

        rerender(
            buildCardRepository(
                'PowerToys',
                'Small educational project',
                'C#',
                67934,
                3816,
                true,
                'https://github.com/microsoft/terminal',
            ),
        );

        expect(favoriteIcon).toHaveClass('fas', 'fa-heart', 'fa-lg');
        expect(favoriteIcon).not.toHaveClass('far');
    });
});
