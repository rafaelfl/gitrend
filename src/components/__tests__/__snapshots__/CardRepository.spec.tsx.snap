// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardRepository component tests it should match snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <button
      class="card__favorite"
    >
      <i
        class="fas fa-heart fa-lg"
      />
    </button>
    <div
      class="card__title"
    >
      <svg
        aria-hidden="true"
        data-view-component="true"
        height="16"
        version="1.1"
        viewBox="0 0 16 16"
        width="16"
      >
        <path
          d="M2 2.5A2.5 2.5 0 014.5 0h8.75a.75.75 0 01.75.75v12.5a.75.75 0 01-.75.75h-2.5a.75.75 0 110-1.5h1.75v-2h-8a1 1 0 00-.714 1.7.75.75 0 01-1.072 1.05A2.495 2.495 0 012 11.5v-9zm10.5-1V9h-8c-.356 0-.694.074-1 .208V2.5a1 1 0 011-1h8zM5 12.25v3.25a.25.25 0 00.4.2l1.45-1.087a.25.25 0 01.3 0L8.6 15.7a.25.25 0 00.4-.2v-3.25a.25.25 0 00-.25-.25h-3.5a.25.25 0 00-.25.25z"
          fill-rule="evenodd"
        />
      </svg>
      <span>
           microsoft / PowerToys
      </span>
    </div>
    <div
      class="card__content"
    >
      Small educational project developed with the objective of presenting the concepts related to the development of backend REST APIs with Node.js for the course Distributed Systems (2021.2) of the Computer Engineering Course - Federal University of M...
    </div>
    <div
      class="card__url"
    >
      <a
        href="https://github.com/microsoft/terminal"
        rel="noreferrer"
        target="_blank"
      >
         https://github.com/microsoft/terminal
      </a>
    </div>
    <div
      class="card__footer"
    >
      <span>
         C#
      </span>
      <span
        class="card__stars"
      >
        <svg
          aria-label="star"
          data-view-component="true"
          height="12"
          role="img"
          version="1.1"
          viewBox="0 0 16 16"
          width="12"
        >
          <path
            d="M8 .25a.75.75 0 01.673.418l1.882 3.815 4.21.612a.75.75 0 01.416 1.279l-3.046 2.97.719 4.192a.75.75 0 01-1.088.791L8 12.347l-3.766 1.98a.75.75 0 01-1.088-.79l.72-4.194L.818 6.374a.75.75 0 01.416-1.28l4.21-.611L7.327.668A.75.75 0 018 .25zm0 2.445L6.615 5.5a.75.75 0 01-.564.41l-3.097.45 2.24 2.184a.75.75 0 01.216.664l-.528 3.084 2.769-1.456a.75.75 0 01.698 0l2.77 1.456-.53-3.084a.75.75 0 01.216-.664l2.24-2.183-3.096-.45a.75.75 0 01-.564-.41L8 2.694v.001z"
            fill-rule="evenodd"
          />
        </svg>
          67,934
      </span>
      <span
        class="card__fork"
      >
        <svg
          aria-label="fork"
          data-view-component="true"
          height="12"
          role="img"
          version="1.1"
          viewBox="0 0 16 16"
          width="12"
        >
          <path
            d="M5 3.25a.75.75 0 11-1.5 0 .75.75 0 011.5 0zm0 2.122a2.25 2.25 0 10-1.5 0v.878A2.25 2.25 0 005.75 8.5h1.5v2.128a2.251 2.251 0 101.5 0V8.5h1.5a2.25 2.25 0 002.25-2.25v-.878a2.25 2.25 0 10-1.5 0v.878a.75.75 0 01-.75.75h-4.5A.75.75 0 015 6.25v-.878zm3.75 7.378a.75.75 0 11-1.5 0 .75.75 0 011.5 0zm3-8.75a.75.75 0 100-1.5.75.75 0 000 1.5z"
            fill-rule="evenodd"
          />
        </svg>
          3,816
      </span>
    </div>
  </div>
</div>
`;
